core/src/main/scala/org/broadinstitute/dsde/rawls/entities/local/LocalEntityProvider.scala

```diff
@@ -282,42 +292,118 @@ class LocalEntityProvider(requestArguments: EntityRequestArguments,
+
+          val metadata: EntityQueryResultMetadata =
+            EntityQueryResultMetadata(unfilteredCount, filteredCount, pageCount)
+          (metadata, entitySource)
+        }
+
+      case _ =>
+        dataSource.inTransaction { dataAccess =>
+          traceDBIOWithParent("loadEntityPage", parentContext) { childContext =>
+            childContext.tracingSpan.foreach { s1 =>
+              s1.putAttribute("pageSize", OpenCensusAttributeValue.longAttributeValue(query.pageSize))
+              s1.putAttribute("page", OpenCensusAttributeValue.longAttributeValue(query.page))
+              s1.putAttribute("filterTerms",
+                              OpenCensusAttributeValue.stringAttributeValue(query.filterTerms.getOrElse(""))
+              )
+              s1.putAttribute("sortField", OpenCensusAttributeValue.stringAttributeValue(query.sortField))
+              s1.putAttribute("sortDirection",
+                              OpenCensusAttributeValue.stringAttributeValue(query.sortDirection.toString)
+              )
+            }
+            for {
+              (unfilteredCount, filteredCount) <- dataAccess.entityQuery.loadEntityPageCounts(workspaceContext,
+                                                                                              entityType,
+                                                                                              query,
+                                                                                              childContext
+              )
+            } yield {
+              val pageCount: Int = Math.ceil(filteredCount.toFloat / query.pageSize).toInt
+              if (filteredCount > 0 && query.page > pageCount) {
+                throw new DataEntityException(
+                  code = StatusCodes.BadRequest,
+                  message = s"requested page ${query.page} is greater than the number of pages $pageCount"
+                )
+              }
+
+              val metadata: EntityQueryResultMetadata =
+                EntityQueryResultMetadata(unfilteredCount, filteredCount, pageCount)
+              val allAttrsStream =
+                dataAccess.entityQuery
+                  .loadEntityPageSource(workspaceContext, entityType, query, childContext)
+                  .transactionally
+                  .withTransactionIsolation(TransactionIsolation.ReadCommitted)
+                  .withStatementParameters(rsType = ResultSetType.ForwardOnly,
+                                           rsConcurrency = ResultSetConcurrency.ReadOnly,
+                                           fetchSize = dataSource.dataAccess.fetchSize
+                  )
+              val dbSource: Source[EntityAndAttributesResult, NotUsed] =
+                Source.fromPublisher(dataSource.database.stream(allAttrsStream))
+
+              val entitySource = EntityStreamingUtils.gatherEntities(dataSource, dbSource)
+
+              (metadata, entitySource)
```

Comment:
This probably works but the structure is confusing. I think there are actually 2 separate transactions:
1. calls `dataAccess.entityQuery.loadEntityPageCounts`
2. calls `dataAccess.entityQuery. loadEntityPageSource`
Structurally, it looks like 2 is within the transaction that includes 1 but it is not. Yet 2 uses information from 1 before 1 completes. I would like to see this broken up a little more cleanly.